Mathlib/Analysis/NormedSpace/PiTensorProduct/InjectiveSeminorm.lean

```diff
@@ -0,0 +1,467 @@
+        LinearEquiv.coe_coe]
+      exact hC.2 F f)
+  map_add' x y := by
+    ext _
+    simp only [map_add, LinearMap.mkContinuousOfExistsBound_apply, LinearMap.coe_comp,
+      Function.comp_apply, ContinuousMultilinearMap.toMultilinearMapLinear_apply,
+      LinearMap.add_apply, LinearMap.flip_apply, LinearEquiv.coe_coe, ContinuousLinearMap.add_apply]
+  map_smul' a x := by
+    ext _
+    simp only [map_smul, LinearMap.mkContinuousOfExistsBound_apply, LinearMap.coe_comp,
+      Function.comp_apply, ContinuousMultilinearMap.toMultilinearMapLinear_apply,
+      LinearMap.smul_apply, LinearMap.flip_apply, LinearEquiv.coe_coe, RingHom.id_apply,
+      ContinuousLinearMap.coe_smul', Pi.smul_apply]
+
+/-- The injective seminorm on `⨂[𝕜] i, Eᵢ`. Morally, it sends `x` in `⨂[𝕜] i, Eᵢ` to the
+`sup` of the operator norms of the `PiTensorProduct.toDualContinuousMultilinearMap x`, for all
+normed vector spaces `F`. In fact, we only take in the same universe as `⨂[𝕜] i, Eᵢ`, and then
+prove in `PiTensorProduct.injectiveSeminorm_bound` that this gives the same result.
+-/
+noncomputable def injectiveSeminorm : Seminorm 𝕜 (⨂[𝕜] i, E i) :=
+  sSup {p | ∃ (G : Type (max (max u_1 u_2) u_3)) (_ : SeminormedAddCommGroup G)
+  (_ : NormedSpace 𝕜 G), p = Seminorm.comp (normSeminorm 𝕜 (ContinuousMultilinearMap 𝕜 E G →L[𝕜] G))
+  (toDualContinuousMultilinearMap (F := G) (𝕜 := 𝕜) (E := E))}
+
+lemma dualSeminorms_bounded : BddAbove {p | ∃ (G : Type (max (max u_1 u_2) u_3))
+    (_ : SeminormedAddCommGroup G) (_ : NormedSpace 𝕜 G),
+    p = Seminorm.comp (normSeminorm 𝕜 (ContinuousMultilinearMap 𝕜 E G →L[𝕜] G))
+    (toDualContinuousMultilinearMap (F := G) (𝕜 := 𝕜) (E := E))} := by
+  rw [Seminorm.bddAbove_iff]
+  set bound : (⨂[𝕜] i, E i) → ℝ :=
+    fun x ↦ Classical.choose (toDualMultilinearMap_bound x)
+  existsi bound
+  rw [mem_upperBounds]
+  intro p hp
+  simp only [Set.mem_image] at hp
+  let ⟨q, hq⟩ := hp
+  simp only [Set.mem_setOf_eq] at hq
+  intro x
+  rw [← hq.2]
+  obtain ⟨⟨G, G₁, ⟨G₂, h⟩⟩⟩ := hq
+  letI := G₁
+  letI := G₂
+  rw [h]
+  simp only [Seminorm.comp_apply, ge_iff_le]
+  let hbound := Classical.choose_spec (toDualMultilinearMap_bound x)
+  exact ContinuousLinearMap.opNorm_le_bound _ hbound.1 (fun f ↦ by
+    simp only [toDualContinuousMultilinearMap_apply_toFun]
+    exact hbound.2 G f)
+
+@[simp]
+theorem injectiveSeminorm_apply (x : ⨂[𝕜] i, E i) :
+    injectiveSeminorm x = ⨆ p : {p | ∃ (G : Type (max (max u_1 u_2) u_3))
+    (_ : SeminormedAddCommGroup G) (_ : NormedSpace 𝕜 G), p = Seminorm.comp (normSeminorm 𝕜
+    (ContinuousMultilinearMap 𝕜 E G →L[𝕜] G))
+    (toDualContinuousMultilinearMap (F := G) (𝕜 := 𝕜) (E := E))}, p.1 x := by
+  refine Seminorm.sSup_apply dualSeminorms_bounded
+
+theorem injectiveSeminorm_bound (f : ContinuousMultilinearMap 𝕜 E F) (x : ⨂[𝕜] i, E i) :
+    ‖lift f.toMultilinearMap x‖ ≤ ‖f‖ * injectiveSeminorm x := by
+  set G := (⨂[𝕜] i, E i) ⧸ LinearMap.ker (lift f.toMultilinearMap)
```

Comment:
When something non trivial happens in a proof, I think it's a good idea to add a few comments about that. Here, you only need to play this game because of universe issues (if I understand correctly), so comments would definitely look worthwhile.

Reply:
Okay, I added a comment explaining what is going on in the proof.